test/upload_validation_test.ts

```diff
@@ -2172,6 +2175,321 @@ describe('Pack metadata Validation', async () => {
+            syncTables: [childTable, parentTable],
+          });
+          await validateJson(metadata);
+          const hierarchy = getSyncTableHierarchy(metadata);
+          assert.deepEqual(hierarchy, {Parent: ['Child']});
+        });
+
+        it('_hasCycle utility works', () => {
+          assert.isTrue(_hasCycle({a: ['b'], b: ['a']}));
+          assert.isTrue(_hasCycle({a: ['a']}));
+          assert.isTrue(_hasCycle({a: ['b'], b: ['c'], c: ['a']}));
+          assert.isTrue(_hasCycle({a: ['c', 'b'], b: ['a'], c: ['d']}));
+          assert.isFalse(_hasCycle({a: ['b'], b: ['c'], c: ['d']}));
+          // This is invalid, but not because it's a cycle
+          assert.isFalse(_hasCycle({a: ['b', 'c'], b: ['c']}));
+        });
+
+        it('does not allow a cycle', async () => {
+          const table1 = makeSyncTable({
+            name: 'Table1',
+            identityName: 'Table1Identity',
+            schema: makeObjectSchema({
+              type: ValueType.Object,
+              id: 'table1prop',
+              primary: 'table1prop',
+              properties: {table1prop: {type: ValueType.String}},
+            }),
+            formula: {
+              name: 'Whatever1',
+              description: '',
+              parameters: [
+                makeParameter({
+                  type: ParameterType.String,
+                  name: 'table2Param',
+                  description: '',
+                  crawlStrategy: {parentTable: {tableName: 'Table2', propertyKey: 'table2prop'}},
+                }),
+              ],
+              async execute() {
+                return {result: []};
+              },
+            },
+          });
+          const table2 = makeSyncTable({
+            name: 'Table2',
+            identityName: 'Table2Identity',
+            schema: makeObjectSchema({
+              type: ValueType.Object,
+              id: 'table2prop',
+              primary: 'table2prop',
+              properties: {table2prop: {type: ValueType.String}},
+            }),
+            formula: {
+              name: 'Whatever3',
+              description: '',
+              parameters: [
+                makeParameter({
+                  type: ParameterType.String,
+                  name: 'table1Param',
+                  description: '',
+                  crawlStrategy: {parentTable: {tableName: 'Table1', propertyKey: 'table1prop'}},
+                }),
+              ],
+              async execute() {
+                return {result: []};
+              },
+            },
+          });
+          const metadata = createFakePack({
+            id: 1013,
+            syncTables: [table2, table1],
+          });
+          await testHelper.willBeRejectedWith(validateJson(metadata), /Sync table parent hierarchy is invalid/);
+          const hierarchy = getSyncTableHierarchy(metadata);
+          assert.isUndefined(hierarchy);
+        });
+
+        it('does not allow multiple parents', async () => {
```

Comment:
I could imagine a world where a child table relies on two parents though? Perhaps none of our data sources have that right now?

Could iterate on this in case that does happen! not blocking

Reply:
Yeah, I reviewed all our 1st party sync tables, and the only ones with multiple required fields were either:
- parent & grandparent
- date ranges
- bools
- enum, defined statically by an autocomplete function

I'm hopeful that we can go a long time without needing to support that.